src/main/scala/exercise2/Stack.scala

```diff
@@ -0,0 +1,60 @@
+package exercise2
+
+import scala.annotation.tailrec
+import scala.util.Try
+
+// FIFO
+sealed trait StackLike[T] {
+  def push(elem: T): StackLike[T]
+
+  // Try because StackLike can be empty => return exception
+  def pop(): Try[StackLike[T]]
+
+  def top(): Option[T]
+
+  def isEmpty: Boolean
+
+  def reverse(): StackLike[T]
+}
+
+/**
+ * The None (tail) situation
+ *
+ * @tparam T
+ */
+case class StackEmpty[T]() extends StackLike[T] {
+  override def push(elem: T): StackLike[T] = Stack(elem, this)
+
+  override def pop(): Try[StackLike[T]]= Try(StackEmpty())
```

Comment:
mm pop on empty stack should be an exception, so `Failure(new RuntimeException("Stack is empty!"))`